src/Compilers/CSharp/Portable/Binder/Binder_Patterns.cs

```diff
@@ -579,10 +579,16 @@ private BoundPattern BindRecursivePattern(RecursivePatternSyntax node, TypeSymbo
             BindPatternDesignation(
                 node.Designation, declTypeWithAnnotations, inputValEscape, typeSyntax, diagnostics,
                 ref hasErrors, out Symbol variableSymbol, out BoundExpression variableAccess);
+            bool isExplicitNotNullTest =
+                node.Designation is null &&
+                boundDeclType is null &&
+                properties.IsDefaultOrEmpty &&
+                deconstructMethod is null &&
+                deconstructionSubpatterns.IsDefaultOrEmpty;
```

Comment:
>deconstructionSubpatterns.IsDefaultOrEmpty [](start = 16, length = 42)

This appears to treat `()` and `() {}` as explicit null tests.  The aren't.  The deconstruction subpatterns need to be null and the properties need to be non-null and empty.